File: problems/Edit Distance/minDistance.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {number}
 */
var minDistance = function(word1, word2) {
    const rows = word1.length;
    const cols = word2.length;
    const dp = Array(1 + rows).fill([]).map(_ => Array(1 + cols).fill(0));

    dp[0] = [...Array(1 + cols).keys()];

    for (let j = 1; rows >= j; ++j) {
        dp[j][0] = j;
    }

    for (let i = 1; rows >= i; ++i) {
        for (let j = 1; cols >= j; ++j) {
            if (word1[i - 1] === word2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1];
            } else {
                dp[i][j] = 1 + Math.min(
                    dp[i - 1][j], // Delete a character
                    dp[i - 1][j - 1], // Replace a character
                    dp[i][j - 1] // Insert a character
                );
            }
        }
    }

    return dp[rows][cols];
};

File: problems/Delete Operation for Two Strings/minDistance.cs
public class Solution {
    public int MinDistance(string word1, string word2) {
        int[,] dp = new int[1 + word1.Length, 1 + word2.Length];

        return word1.Length + word2.Length - 2 * lcs(word1, word2, word1.Length, word2.Length, dp);
    }

    public int lcs(String s1, String s2, int m, int n, int[,] dp) {
        if (m == 0 || n == 0) {
            return 0;
        }

        if (dp[m, n] > 0) {
            return dp[m, n];
        }

        if (s1[m - 1] == s2[n - 1]) {
            dp[m, n] = 1 + lcs(s1, s2, m - 1, n - 1, dp);
        } else {
            dp[m, n] = Math.Max(lcs(s1, s2, m, n - 1, dp), lcs(s1, s2, m - 1, n, dp));
        }

        return dp[m, n];
    }
}
